src/client/credentials_json_client.ts

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+import { createSign } from 'crypto';
+import { AuthClient } from './auth_client';
+import { toBase64, fromBase64, writeSecureFile } from '../utils';
+
+/**
+ * Available options to create the WorkloadIdentity client.
+ *
+ * @param projectID User-supplied value for project ID. If not provided, the
+ * project ID is extracted from the credentials JSON.
+ * @param credentialsJSON Raw JSON credentials blob.
+ */
+interface CredentialsJSONClientOptions {
+  projectID?: string;
+  credentialsJSON: string;
+}
+
+export class CredentialsJSONClient implements AuthClient {
+  readonly #projectID: string;
+  readonly #credentials: Record<string, string>;
+
+  constructor(opts: CredentialsJSONClientOptions) {
+    this.#credentials = this.parseServiceAccountKeyJSON(opts.credentialsJSON);
+    this.#projectID = opts.projectID || this.#credentials['project_id'];
+  }
+
+  /**
+   * parseServiceAccountKeyJSON attempts to parse the given string as a service
+   * account key JSON. It handles if the string is base64-encoded.
+   */
+  parseServiceAccountKeyJSON(str: string): Record<string, string> {
+    if (!str) {
+      return {};
+    }
+
+    str = str.trim();
+
+    // If the string doesn't start with a JSON object character, it is probably
+    // base64-encoded.
+    if (!str.startsWith('{')) {
+      str = fromBase64(str);
+    }
+
+    try {
+      return JSON.parse(str);
+    } catch (e) {
+      throw new SyntaxError(`Failed to parse credentials as JSON: ${e}`);
+    }
+  }
+
+  /**
+   * getAuthToken generates a token capable of calling the iamcredentials API.
+   */
+  async getAuthToken(): Promise<string> {
+    const header = {
+      alg: 'RS256',
+      typ: 'JWT',
+      kid: this.#credentials['private_key_id'],
+    };
+
+    const now = Math.floor(new Date().getTime() / 1000);
+
+    const body = {
+      iss: this.#credentials['client_email'],
+      sub: this.#credentials['client_email'],
+      aud: 'https://iamcredentials.googleapis.com/',
+      iat: now,
+      exp: now + 3599,
```

Comment:
I didnt get a chance to try this in my PR but cant we modify this value to control lifetime? When I was looking at the spec previously it seemed like this can support being extended upto 12h with `allowServiceAccountCredentialLifetimeExtension`  and the body can support scopes too.

Reply:
Yes and no. This is a confusing area, so let me try to explain.

What I did in this PR was "normalize" the WIF and JSON auth clients and move OAuth and ID Token minting into a shared client. The `iamcredentials` API accepts 3 types of authentication:

- OAuth 2.0
- JWT
- Federated tokens (from WIF)

For the WIF client, `getAuthToken()` does the sts WIF dance up to the federated token and returns the federated token. For the JSON client, `getAuthToken()` uses the service account's private key to mint a signed JWT scoped to the `iamcredentials` API (L103). That token is _never_ given back to the user (it's not exposed as an output in the Action). Instead, it's used the same as the federated token when calling the `iamcredentials` API.

It is not possible to extend this expiration beyond 1hr, even if the org policy lifetime extension is in place. Similarly, you can't mint an OAuth 2.0 token using the oauth2 endpoints > 1h even if the org policy is set. That org policy (and the feature) is only available via the `iamcredentials` API.

If a customer has enabled the `allowServiceAccountCredentialLifetimeExtension` org policy, they can set `access_token_lifetime` to a value greater than 1hr when `token_format` is `"access_token"`. Using the `iamcredentials` API is the only way to get tokens that expires > 1hr via that org policy.

Reply:
This makes sense now! Thanks for the detailed explanation.